File: AAD/p2.py
import heapq

class Node:
    def __init__(self, data, freq, left=None, right=None):
        self.data=data
        self.freq=freq
        self.left=left
        self.right=right
    
    def __lt__(self, other):
        return self.freq<other.freq

    def __del__(self):
        del self.left
        del self.right

def print_code(root, str):
    if not root:
        return 
    
    if not root.left and not root.right:
        print(f"{root.data} {str} Required Bits: {root.freq * len(str)}")
        return 

    print_code(root.left, str+'0')
    print_code(root.right, str+'1')

def haffman_coding(char_freq):
    queue=[Node(data, freq) for data, freq in char_freq]
    heapq.heapify(queue)

    while len(queue)>1:
        left=heapq.heappop(queue)
        right=heapq.heappop(queue)

        merged=Node('$', left.freq + right.freq, left, right)
        heapq.heappush(queue, merged)

    print("\nCharacter Code: ")
    print_code(queue[0],"")


if __name__== "__main__":

    num= int(input("Enter the number of characters: "))
    char_freq=[]
    for i in range (num):
        c=input("Enter the character: ")
        freq=int(input("Enter the frequency of character: "))
        char_freq.append((c, freq))
    
    haffman_coding(char_freq)

File: AAD/p2.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <memory> // For std::unique_ptr
using namespace std;

class Node
{
public:
    char data;
    int freq;
    Node *left, *right;

    Node(char data, int freq, Node *left = nullptr, Node *right = nullptr)
    {
        this->data = data;
        this->freq = freq;
        this->left = left;
        this->right = right;
    }

    // Destructor to free left and right nodes recursively
    ~Node()
    {
        delete left;
        delete right;
    }
};

// Comparator for the priority queue
struct Compare
{
    bool operator()(Node *a, Node *b)
    {
        return a->freq > b->freq; // Min-heap based on frequency
    }
};

// Recursive function to print Huffman codes
void printCode(Node *head, string code)
{
    if (!head)
        return;

    // If this is a leaf node, print the character and its code
    if (!head->left && !head->right)
    {
        cout << head->data << "  " << code << "  "
             << "Required Bits: " << head->freq * code.length() << endl;
        return;
    }

    // Recur for left and right children
    printCode(head->left, code + '0');
    printCode(head->right, code + '1');
}

int main()
{
    int n;
    cout << "Enter the number of characters: ";
    cin >> n;

    if (n <= 0)
    {
        cout << "The number of characters must be positive." << endl;
        return 1;
    }

    char c;
    int freq;
    priority_queue<Node *, vector<Node *>, Compare> queue;

    for (int i = 0; i < n; i++)
    {
        cout << "Enter the character: ";
        cin >> c;
        cout<< endl;
        cout << "Enter the frequency of " << c << ": ";
        cin >> freq;

        if (freq < 0)
        {
            cout << "Frequency cannot be negative." << endl;
            return 1;
        }

        queue.push(new Node(c, freq));
    }

    // Build the Huffman Tree
    while (queue.size() > 1)
    {
        Node *left = queue.top();
        queue.pop();
        Node *right = queue.top();
        queue.pop();

        // Internal node with frequency equal to the sum of the two child nodes
        queue.push(new Node('$', left->freq + right->freq, left, right));
    }

    cout << "\nCharacter Codes:\n";
    printCode(queue.top(), "");

    // Free the memory allocated for the Huffman Tree
    delete queue.top();

    return 0;
}
